Consolidate Triton cache into Inductor cache

oulgen opened the pull request:
Summary:
This diff/PR attempts to consolidate Triton caching into the Inductor caching so that there can be just one cache that unifies them both, reducing network requests and increasing success rate.

Implementation details can be found via reading the code or the post: https://fb.workplace.com/groups/1553867532149891/posts/1605037517032892

I did not use the Autotune bundler code at all since I want to simplify that and merge it into this on the next diff/PR.

In terms of instrumentation
1) Dynamo compile: `triton_bundler_time_saved_s` this is sum of all triton.compile calls. We dont have to use the specific number, can use this as a binary value.
2) Events table: I used dynamo_timed to measure how much time we spend on bundler collect and write functions which is all the work we do in this diff
3) TLParse: I emitted number of kernels and triton_bundler_time_saved_s into tlparse as well

Test Plan: Updated unit tests

Adhoc running
```
TORCHINDUCTOR_BUNDLE_TRITON_INTO_FX_GRAPH_CACHE=1 buck2 run @mode/opt //scripts/oulgen:runner
```
gives
https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmpmTZt6b/0_0_0/fx_graph_cache_hit_4.json
<img width="771" alt="image" src="https://github.com/user-attachments/assets/478782a2-ee47-40cb-b723-fcac2bf9dd93">




Differential Revision: D64504909




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec

oulgen commented on torch/_inductor/runtime/triton_bundler.py:
```diff
@@ -0,0 +1,127 @@
+                directory = os.path.join(directory, str(artifacts.device))
+            Path(directory).mkdir(parents=True, exist_ok=True)
+
+            for artifact in artifacts.artifacts:
+                filepath = os.path.join(directory, artifact.filename)
+                with open(filepath, "wb") as file:
+                    file.write(artifact.payload)
+                counters["inductor"]["triton_bundler_write_kernel"] += 1
```
Does this write need to be atomic? Or is this safe because we are the only ones writing at this point in code?

oulgen commented:
@ezyang I'm not really sure how to integrate this into TLParse.. Any recommendations?

ezyang commented:
Hi, I don't see any logging / trace structured updates / scuba updates, can you think about what kind of instrumentation you want to add and propose something here?

ezyang commented on torch/_inductor/runtime/triton_bundler.py:
```diff
@@ -0,0 +1,127 @@
+    @classmethod
+    def collect(cls) -> List[TritonKernelArtifacts]:
+        if not TritonBundler.is_enabled():
+            cls._entries = None
```
I am a bit confused about thread safety here. You are talking about `_entries` writes OK from multiple threads to be atomic. Is collect a sync point then or something?

oulgen replied:
I'm probably being overly cautious here. collect is executed when we are writing the inductor cache entry to disk/remote, so at this point we expect everything to have completed and this is true by design (no need for any sync primitive).

c00w commented on torch/_inductor/runtime/triton_bundler.py:
```diff
@@ -0,0 +1,127 @@
+class TritonBundler:
+    # It is safe for multiple threads to insert entries as insert operation
+    # of dict is atomic
+    _entries: Optional[Dict[TritonBundleEntry, None]] = None
```
Nit - is this normal pytorch style? or should we just add a lock to make this thread safety explicit?

oulgen replied:
The operation is atomic, why would we need a lock?

c00w replied:
Documentation. (also do you think the overhead will be significant here? If you do we can drop it, but assuming what 100k triton compiles, and 100ns for a mutex acquisition, you get 10ms in overhead).

c00w commented on torch/_inductor/runtime/triton_bundler.py:
```diff
@@ -0,0 +1,127 @@
+                        assert os.path.isfile(filepath)
+                        with open(filepath, "rb") as file:
+                            artifacts.append(
+                                TritonKernelArtifact(filename, file.read())
```
qq - Is there a reason we don't want to keep the triton items in memory? Reading them off the file system is fine, but it does introduce more points of failure (i.e.disk space is full, multiple processes accidentally writing at the same time, etc...).

oulgen replied:
We still have to rely on the triton compiler, so this is a way to feed it off of to the triton compiler. We do not have a deep integration with the compiler to feed it in memory objects

c00w replied:
sgtm, maybe mention that in a comment?

c00w commented on torch/_inductor/runtime/triton_bundler.py:
```diff
@@ -0,0 +1,127 @@
+
+            for artifact in artifacts.artifacts:
+                filepath = os.path.join(directory, artifact.filename)
+                with open(filepath, "wb") as file:
```
This isn't really safe. The normal way to do this is to write to a temporary file (either in the same directory or elsewhere), and then swap them, since that's a atomic operation on most file systems. It also helps handle crashes with partial writes etc....

oulgen replied:
I dont think we need this to be atomic, which is what i was asking below, but you're right about crashes, so i'll update it to use temp file and swap

c00w commented on torch/_inductor/runtime/triton_bundler.py:
```diff
@@ -0,0 +1,127 @@
+            for artifact in artifacts.artifacts:
+                filepath = os.path.join(directory, artifact.filename)
+                with open(filepath, "wb") as file:
+                    file.write(artifact.payload)
```
Is there a hash or something we can use here to help detect corruption? or even a version number if we ever want to update the on disk format?

c00w replied:
I see a kernel_hash above, not sure how easy that is to check.

oulgen replied:
kernel_hash is just the folder name triton decided, i dont think it is complete hash. We should sha256 it if you want when we save it if you want but we havent done that anywhere else before

c00w replied:
Yeah - I'm probably just paranoid, but if we ever have corruption issues, it'll help us figure it out. Your call.

c00w commented on torch/_inductor/runtime/triton_bundler.py:
```diff
@@ -0,0 +1,127 @@
+            cls._entries = None
+            return []
+
+        if (entries := cls._entries) is not None:
```
Should we assert that cls._entries it not None here? The begin compile method seems to indicate that we're trying to ensure one compile at a time, but we don't check that any entries are written here.

Or are we instead going to sometimes have to _entries even for valid compiles so we can't assert correctness here?

oulgen replied:
I had this it this way because what if someone enables just knobs in the middle of execution, so entries is None but collect operation executes because it was turned on midway

c00w replied:
Hmmm - I guess I'm curious about if it makes sense to have begin_compile around at all? just set _entires in put, read + clear them in collect?  But very much your call.

Seperately if JK is turned on between begin_compile + put, we're going to not write to the filesystem during put, not sure if that's an issue (if inductor is going to read the files later).

c00w commented on test/inductor/test_codecache.py:
```diff
@@ -140,30 +142,42 @@ def fn(x, y):
     @parametrize("dynamic", (False, True))
-    def test_remote_cache_load_function(self, device, dtype, dynamic):
+    @parametrize("bundle_triton", (False, True))
+    def test_remote_cache_load_function(self, device, dtype, dynamic, bundle_triton):
```
Is it possible to test doing a compile in two threads at the same time to make sure there are no conflicts? Or do we expect that's explicitly non supported.

(But maybe two processes at the same time are?)

oulgen replied:
I dont think we ever compile in two threads? Only the triton compilation could happen in subprocs and that's already enabled by default, so it should be running on this

c00w commented on test/inductor/test_codecache.py:
```diff
@@ -1,6 +1,7 @@
 # Owner(s): ["module: inductor"]
```
Style question - Is it pytorch style to not write explicit tests for the trion_bundler and maybe triton_heuristics files?

oulgen replied:
All the tests I have seen so far have been end to end tests with counters to verify data. I could write more if you want but I'm just adhering to what currently exists

c00w replied:
Could we test the corner and errors cases in triton_bundler? I.e. begin_compile called twice, collect called twice, disk reads + writes throw errors?

Maybe even writes from two threads to make sure stuff is thread safe?

c00w commented:
re: data sources, at a minimum can we log whether this confis is enabled into dynamo_compile and pt2_remote_cache?

c00w commented on torch/_inductor/runtime/triton_bundler.py:
```diff
@@ -0,0 +1,127 @@
+            custom_dir = False
+            basedir = os.path.join(cache_dir(), "triton")
+
+        for artifacts in bundle:
```
worth logging (counter or something) how many bytes we're writing into the cache?

c00w commented on torch/_inductor/codecache.py:
```diff
@@ -1119,6 +1120,9 @@ def iterate_over_candidates() -> Generator[CompiledFxGraph, None, None]:
             write_atomic(artifact_path, code, make_dirs=True)
 
+        if bundle := graph._triton_bundle:
+            TritonBundler.write_bundle_to_file_system(bundle)
```
qq - Worth recording how long this takes in case the disk is slow?

c00w commented on torch/_inductor/codecache.py:
```diff
@@ -1464,6 +1468,7 @@ def load(  # type: ignore[no-untyped-def]
             compiled_graph._time_taken_ns = time_ns() - start_time
             cache_key = key_info[0]
             compiled_graph._fx_graph_cache_key = cache_key
+            compiled_graph._triton_bundle = TritonBundler.collect()
```
qq - Worth recording how long this takes?

ezyang commented on torch/_inductor/runtime/triton_bundler.py:
```diff
@@ -0,0 +1,127 @@
+from pathlib import Path
+from typing import Dict, List, Optional
+
+import torch
```
@jansel When do I put a file in torch._inductor.runtime as opposed to somewhere else?

jansel replied:
Our generated code should only import from torch._inductor.runtime.

ezyang commented on torch/_inductor/runtime/triton_bundler.py:
```diff
@@ -0,0 +1,127 @@
+    @classmethod
+    def put(cls, kernel_hash: str, device: int) -> None:
+        if (entries := cls._entries) is not None:
+            # CachingAutotuner.__init__ unconditionally sets TRITON_CACHE_DIR
```
This is not really your problem but I want to point out that setenv on TRITON_CACHE_DIR is not thread safe lol

ezyang commented on torch/_inductor/runtime/triton_bundler.py:
```diff
@@ -0,0 +1,127 @@
+log = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass(frozen=True)
```
Some minimal documentation for these dataclasses would be helpful on review

ezyang commented on torch/_inductor/runtime/triton_bundler.py:
```diff
@@ -0,0 +1,127 @@
+class TritonBundler:
+    # It is safe for multiple threads to insert entries as insert operation
+    # of dict is atomic
+    _entries: Optional[Dict[TritonBundleEntry, None]] = None
```
Is this... just an OrderedSet? lol

ezyang commented:
Despite commenting a bit, I might not be a good reviewer for this, there's a lot of indirect threading in and out of the Triton backend that I am not familiar with the call sites for. I'll work harder to review it if Bert/Niko don't pick up the review.

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -812,6 +825,7 @@ class BwdCompilationMetrics:
     fail_type: Optional[str]
     fail_reason: Optional[str]
     remote_cache_time_saved_s: Optional[float]
+    triton_bundler_time_saved_s: Optional[float]
```
Our new guidance is to save these as int and ms

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,188 @@
+    """
+    When we have compiled a triton kernel, we take note of that kernel by
+    its triton generated hash, its device, and where this kernel is located.
+    This is the minimum information we can use to later retreive this kernel
```
retrieve

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,188 @@
+    @classmethod
+    def collect(
+        cls,
+    ) -> Tuple[List[TritonKernelArtifacts], Optional[TritonBundlerMetadata]]:
```
docs please! (and on all the other methods)

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,188 @@
+            return [], None
+
+        with dynamo_timed(
+            key="TritonBundler::collect", phase_name="TritonBundler", fwd_only=False
```
why colons in the key?

ezyang replied:
@jamesjwu I'm not convinced these should have phase name. It's good to have these available in the detailed table but it's not clear I want two new dedicated columns in dynamo_compile for these timings only. Especially because these are filesystem interaction only.

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,188 @@
+            return [], None
+
+    @staticmethod
+    def write(bundle: List[TritonKernelArtifacts]) -> Optional[TritonBundlerMetadata]:
```
This is cache read!

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,188 @@
+        if not TritonBundler.is_enabled():
+            return
+        assert cls._entries is None
+        cls._entries = OrderedSet()
```
collect ends compile

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,188 @@
+        cls._entries = OrderedSet()
+
+    @classmethod
+    def put(cls, kernel_hash: str, compile_time_ns: int, device: int) -> None:
```
If it's not too much knock on effect, have this pass around ms on int.

ezyang replied:
Q: Will this be called from multiple processes in parallel compilation? @oulgen to work it out

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,188 @@
+
+    # It is safe for multiple threads to insert entries as insert operation
+    # of OrderedSet which uses dict underhood is atomic
+    _entries: Optional[OrderedSet[TritonBundleEntry]] = None
```
Does this really need to be a set?

masnesral commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -454,10 +455,14 @@ def _precompile_config(self, cfg: Config, warm_cache_only: bool):
-            return (
-                triton.compile(*compile_args, **compile_kwargs),
-                None,
+            start_time = time.time_ns()
```
So actually, IIUC warm_cache_only is only set for the case where we compile in a subprocess, i.e., to warm the cache. So I think this will effectively update the TritonBundler instance in the subprocess and is therefore useless? For reasons I didn't dig into, there are clearly cases where we do more work below, but that only happens in the main process. So the `TritonBundler.put` at the end of this method looks like the important one.

But is there a reason not to call `TritonBundler.put` in the caller (precompile), and only for the case where `warm_cache_only == False`? That would seem to be a more natural location to me.

oulgen replied:
There are quite a few precompile calls whereas triton.compile calls are very localized in this function. Moreover i only want to measure time taken in triton.compile and not rest of this function

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,188 @@
+        with dynamo_timed(
+            key="TritonBundler::collect", phase_name="TritonBundler", fwd_only=False
+        ):
+            if (entries := cls._entries) is not None:
```
nit: you don't need a walrus here, you can just pre-assign it lol

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,188 @@
+                            assert os.path.isfile(filepath)
+                            with open(filepath, "rb") as file:
+                                artifacts.append(
+                                    TritonKernelArtifact(filename, file.read())
```
The bare file.read() here makes me nervous but I guess these files are not too big

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,188 @@
+                                artifacts,
+                            )
+                        )
+                        total_compile_time_ns += entry.compile_time_ns
```
This potentially overrepresents end-to-end compile time because when parallel compilation is enabled the true wall time is less than the sum of all the individual compile times of each artifact

oulgen replied:
What would you recommend here? doing max based on whether parallel compile is enabled or not?

ezyang replied:
But you're not going to have infinite parallelism right. So max isn't right either.

It also gets murkier because you're going to use this to adjust compiler collective right? So now here's another asymmetric situation to worry about: a bundled cache is generated when parallel compile is on, let's suppose perfect parallelism so 20 kernels x 10s = 10s total.  But someone else only partially cache hits in a setting where parallel compile is off, so now you have 20 x 10s = 200s on the other ranks. NCCL timeout!

For NCCL timeout, it seems safe to estimate the *upper bound* of how much compile time is potentially saved. So summing is the right quantity to use for NCCL timeout. But for "time saved" analysis, you would rather have something more realistic that accounts for parallel compile, or need lots of disclaimers on the data.

oulgen replied:
This is actually worse. Remember how @masnesral and I were explaining that we compile kernels twice once in subprocess and once in mainprocess. SO this number is the second compilation, not first.

But this is actually fine for NCCL since we are not using this number to adjust, we are using compile_fx time (already committed) since that includes this timing end-to-end. With a partial cache hit, it does get muddier though, but we are arleady fuding the number to be higher anyway

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,188 @@
+                        with open(filepath, "wb") as file:
+                            file.write(artifact.payload)
+                        counters["inductor"]["triton_bundler_write_kernel"] += 1
+                    shutil.copytree(tmp_dir, directory, dirs_exist_ok=True)
```
... why a copytree?

oulgen replied:
is there a better way to copy a folder?

ezyang replied:
Naively I would have expected you to write the files directly into the locations they were expected to be (or, if atomicity is needed, rename them in), not sure why we write it one place and then do a copy

oulgen replied:
That's what I originally did, but Colin asked me to change it since what if we fail midway writing files, we want all or nothing behavior

ezyang replied:
You can still do that though, you want to do an atomic file rename. In fact, copytree is implemented under the hood by writing out 16KB chunks https://stackoverflow.com/questions/20873723/is-pythons-shutil-copyfile-atomic so it's not even atomic!

oulgen replied:
@c00w ?

ezyang commented on torch/_inductor/graph.py:
```diff
@@ -1946,6 +1947,7 @@ def _compile_to_module(self) -> ModuleType:
 
         inductor_meta = autotune_cache.inductor_meta_from_config()
         AutotuneCacheBundler.begin_compile(inductor_meta, code=code)
+        TritonBundler.begin_compile()
```
Kind of wondering why this doesn't live in load_by_key_path (which is what actually triggers Triton compiles)

oulgen replied:
load_by_key_path is called right after this. Do you want this to be inside load_by_key_path? The other caller of load_by_key_path in codecache does not need this as we are not going to be writing to cache after that

ezyang replied:
Well, I'm on the record for saying that load_by_key_path being called from two sites (once from here, once from the cache hit site) has been the source of endless confusion (e.g., Triton compile time on cache hit getting incorrectly attributed to inductor phase instead of code_gen phase). I guess if you want @masnesral to fix it all in one go that's fine. Not sure if Sam has an opinion.

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,188 @@
+            custom_dir = True
+            if basedir is None:
+                custom_dir = False
+                basedir = os.path.join(cache_dir(), "triton")
```
This feels potentially duped with preexisting code somewhere? Is it possible to DRY this?

ezyang replied:
There is a potentially silent performance problem that can happen when there bugs in this code, which goes like this: we accidentally put the Triton cache products in not quite the right place. Because there's no way to say "Triton, you MUST cache hit, error otherwise", Triton doesn't see the cache products and then goes and ahead and recompiles. This is completely silent and you only notice it because you are cache hit but it's still taking a long time to do Triton compilation.

The most obvious way to fix this is to somehow run Triton in a mode where cache hit is mandatory and error if it doesn't. But I don't know if we actually have an interposition point to do this (might need Triton change.) If we can't make it an error, then the other thing we can do is make sure we are carefully measuring compile time when we actually run Triton, and we can look for outliers WHEN bundled cache hit (it should be some low number and any time it is high there is a problem.) Unclear what logging you can have to further diagnose a problem like this though!

oulgen replied:
> Is it possible to DRY this?

What does this mean?

Regarding changing Triton, it is something we can try negotiating with OAI, but likely will take a while

ezyang replied:
DRY = Don't Repeat Yourself

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,188 @@
+                if not custom_dir:
+                    # When we do not use a custom triton cache dir, we separate kernels
+                    # by their device
+                    directory = os.path.join(directory, str(artifacts.device))
```
Ditto here, feels like it should be DRY

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,188 @@
+        from torch._inductor import config
+
+        if (b := config.bundle_triton_into_fx_graph_cache) is not None:
+            return b
```
Don't you have to respect disable all caches config too?

oulgen replied:
If inductor cache is disabled, this code will never run, so we dont need to check here

ezyang replied:
This doesn't seem obviously true? begin_compile is called from graph.py and will then hit this, and I'm pretty sure that code gets called always

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,188 @@
+            return b
+
+        if not config.is_fbcode():
+            return False
```
How come? This seems useful for OSS too.

oulgen replied:
OSS Uses the config option above, this check is guarding the justknobs_check check

ezyang replied:
JK checks don't need to be is_fbcode guarded. But OK, fine, this will all get ripped out when @c00w's PR lands

c00w replied:
+1 don't need to guard JK check on fbcode (it just always returns true in OSS).

oulgen replied:
But in this case, we dont want it to default to be on in oss, so i think lets keep this until the reworked configs

ezyang commented on torch/_inductor/codecache.py:
```diff
@@ -1133,6 +1136,16 @@ def iterate_over_candidates() -> Generator[CompiledFxGraph, None, None]:
+                    meta.total_compile_time_ns,
+                    is_backward=is_backward,
+                    triton_bundler=True,
+                )
```
You're doing this completely symmetrically to the old column but we know there are problems with the old column so I'd like to do this correctly from the get go.

So what exactly do we want out of this field. I want to know if I had a bundled cache hit or not. So that seems like the bare minimum instrumentation to have. Might be useful to know how many kernels / how big the kernels I hit on in the bundle are. Basically, anything you would have logged to the pt2_remote_cache table, log it here instead, with appropriate aggregation in case there are multiple Inductor compiles in the dynamo_compile span.

Now, how about the actual timings? I think I am interested in the timing that I put into the NCCL ephemeral timeout, because I may want to specifically debug problems where this timeout number is too large. I don't think I'm interested in imputed savings here though (which is what the current field naming suggests). There are too many caveats (parallel versus sequential, also our recent findings about DtoH syncs in compiler causing unpredictable interactions with the in flight eager operations) that I really only want to rely on natural A/B test for sizing the impact here to start.

So concretely, don't call this remote cache time saved. Call it something like nccl ephemeral timeout added or something.

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -224,7 +224,11 @@ def _add_time_spent(key: str, phase_name: str, time_spent: float) -> None:
+# triton_bundler argument is used to distinguish between whole inductor cache or
+# just partial triton_bundler
+def add_remote_cache_time_saved(
+    time_saved_ns: int, is_backward: bool, triton_bundler: bool
```
nit: make the bool args kwarg only

ezyang commented on torch/_inductor/codecache.py:
```diff
@@ -1478,6 +1494,9 @@ def load(  # type: ignore[no-untyped-def]
             compiled_graph._fx_graph_cache_key = cache_key
+            compiled_graph._triton_bundle, triton_bundler_meta = TritonBundler.collect()
+            if (meta := triton_bundler_meta) is not None:
+                cache_info["triton_bundler_meta"] = str(meta)
```
nit: the walrus here saves you like... five characters lol

ezyang commented on test/inductor/test_codecache.py:
```diff
@@ -1,6 +1,7 @@
 # Owner(s): ["module: inductor"]
 import os
```
I need to think about what I want out of unit tests more, but I think running this on an end to end model (even the OSS benchmark suite) with Triton caching turned off and showing that it results in end-to-end speedup over Inductor caching only would be a good part of the test plan.

ezyang commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -454,10 +455,14 @@ def _precompile_config(self, cfg: Config, warm_cache_only: bool):
-                triton.compile(*compile_args, **compile_kwargs),
-                None,
+            start_time = time.time_ns()
+            binary = triton.compile(*compile_args, **compile_kwargs)
```
Blocking I guess.

ezyang commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -470,7 +475,9 @@ def _precompile_config(self, cfg: Config, warm_cache_only: bool):
             try:
+                start_time = time.time_ns()
                 binary = triton.compile(*compile_args, **compile_kwargs)
+                compile_time_ns = time.time_ns() - start_time
```
Because you don't report these stats, except incorporating them into cache products, it is impossible to observe the cache hit performance of triton.compile (which is surely nonzero, because it has to go actually compute cache keys and then do the lookup.)

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,217 @@
+                        # Just append one of them without the extension
+                        kernel_names.append(Path(artifact.filename).stem)
+                # Atomic on POSIX systems
+                os.replace(tmp_dir, directory)
```
A little confused here:

> os.replace(src, dst, *, src_dir_fd=None, dst_dir_fd=None)
> Rename the file or directory src to dst. If dst is a non-empty directory, OSError will be raised.

You seem to pass a directory on the RHS, this doesn't seem like it could possibly work?

oulgen replied:
we chatted offline, this does work

ezyang reviewed (changes requested):
New os.replace logic is not quite right

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,223 @@
+
+                if os.path.exists(directory) and len(os.listdir(directory)) != 0:
+                    # If directory already exists, we bail out and leave
+                    # local disk to take care of caching
```
I need more description here, specifically, where are documenting the fact that we assume exclusive access to the directory?

oulgen committed:
Consolidate Triton cache into Inductor cache (#138239)

Summary:

This diff/PR attempts to consolidate Triton caching into the Inductor caching so that there can be just one cache that unifies them both, reducing network requests and increasing success rate.

Implementation details can be found via reading the code or the post: https://fb.workplace.com/groups/1553867532149891/posts/1605037517032892

I did not use the Autotune bundler code at all since I want to simplify that and merge it into this on the next diff/PR.

In terms of instrumentation
1) Dynamo compile: `triton_bundler_time_saved_s` this is sum of all triton.compile calls. We dont have to use the specific number, can use this as a binary value.
2) Events table: I used dynamo_timed to measure how much time we spend on bundler collect and write functions which is all the work we do in this diff
3) TLParse: I emitted number of kernels and triton_bundler_time_saved_s into tlparse as well

Test Plan:
Updated unit tests

```
TORCHINDUCTOR_BUNDLE_TRITON_INTO_FX_GRAPH_CACHE=1 buck2 run mode/opt //scripts/oulgen:runner
```
https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmpmTZt6b/0_0_0/fx_graph_cache_hit_4.json
 {F1944553347}

Reviewed By: ezyang

Differential Revision: D64504909

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,234 @@
+
+    # __grp__kernel_name.json contains metadata with source code paths
+    # we use this as sentinal value for search and replace
+    _REPLACE_BYTES: bytes = b"[REPLACE]"
```
There will be some terrifying SEV when this mucks over some unrelated byte string that just *happens* to be `[REPLACE]`. One way to protect against this is to assert that there are no unrelated occurrences of this string before we munge file paths.

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,234 @@
+                    path = os.path.join(entry.directory, entry.kernel_hash)
+                    if not os.path.exists(path):
+                        continue
+                    for filename in os.listdir(path):
```
So post our discussion, we have exclusive access to the kernel directory at this point in time (no one potentially racing us writing files into this directory while we are collecting).

oulgen replied:
i added this to the doc block above

ezyang commented on torch/_inductor/triton_bundler.py:
```diff
@@ -0,0 +1,234 @@
+                # Random ID to avoid any collisions
+                rnd_id = str(uuid.uuid4())
+                tmp_dir = os.path.join(basedir, f"tmp.{rnd_id}")
+                os.makedirs(tmp_dir)
```
nit: uuid to generate tempfile is not actually the recommended way. On POSIX, you can use mkstemp or similar to create a temporary file in a directory that is guaranteed not to conflict with any other file.

ezyang replied:
But actually, because you have exclusive access, you don't even need a random ID, you can just write to a well known temp file name and because you assume no one else is mucking with this dir it's fine. You only need to guard against premature termination

ezyang replied:
although you do have to delete stale tmp dir if there are any lying around :P If it's a well known name, it's easier to make sure that you cleanup the next time around.

ezyang reviewed (approved):
None of the comments are blocking
